### PR TITLE
Add recipe for tesseract

### LINKED ISSUE #4647: Package tesseract
https://github.com/tesseract-ocr/tesseract

Note that there is a bioconda package for this: https://anaconda.org/bioconda/tesseract

### PR DESCRIPTION
**Pull request by CurtLH**
Closes #4647

### CODE REVIEW

**Comment by CurtLH** [comment]
@conda-forge/staged-recipes -- can you provide any guidance on how to handle recipes that are in bioconda but not conda-forge?  Is this something where a recipe will never be merged if it's already in bioconda?

**Comment by CurtLH** [comment]
This recipe was largely inspired by the existing [`tesseract recipe`](https://github.com/bioconda/bioconda-recipes/blob/1f9ec5890d7d67c075a134906796ffc0962ceeae/recipes/tesseract/3.04.01/meta.yaml) in bioconda.

**Comment by scopatz** [comment]
>  Is this something where a recipe will never be merged if it's already in bioconda?

Actually many bioconda recipes are in conda and there is no problem duplicating packages here.

**Comment by scopatz** [comment]
I don't know that I agree with what the linter is doing here.  Perhaps @conda-forge/bioconda-recipes can shed some light.

File: recipes/leptonica/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    - {{ compiler('c') }}
+  host:
+    - cmake
+    - gcc       # [linux]
+    - libwebp 
+  run:
+    - zlib
+    - jpeg
+    - libpng
+    - libtiff
```
Comment (jakirkham) [inline recipes/leptonica/meta.yaml]: These likely all need to be in `host` as well. C/C++ code compiles libraries that link to libraries. So they have to be around during the build for the linking part to occur. They also need to be included at run time with matching versions (we take care of that for you under the hood). This differs from Python where they just need to be around whenever an `import` line calls for them. Hope that helps.

File: recipes/leptonica/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - cmake
+    - gcc       # [linux]
+    - libwebp 
+  run:
+    - zlib
+    - jpeg
```
Comment (jakirkham) [inline recipes/leptonica/meta.yaml]: Looks like this needs `jpeg` and `openjpeg`. So would add both to `host` and `run`.

File: recipes/leptonica/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+build:
+  number: 0
+  skip: True  # [osx or win]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - cmake
```
Comment (jakirkham) [inline recipes/leptonica/meta.yaml]: This is using `pkg-config` as well. Currently it is reaching for the system copy, but we can provide our own (on macOS too). So please add it here.

Based on these log lines.

```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.23") 
```

ref: https://circleci.com/gh/conda-forge/staged-recipes/27445

File: recipes/leptonica/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    - gcc       # [linux]
```
Comment (jakirkham) [inline recipes/leptonica/meta.yaml]: This is handled by the `compiler` portion above. So it can be dropped.

File: recipes/leptonica/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    - libwebp
```
Comment (jakirkham) [inline recipes/leptonica/meta.yaml]: This should be in `run` as well.

File: recipes/tesseract/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [osx or win]
+
+requirements:
+  host:
+    - gcc       # [linux]
+    - llvm      # [osx]
```
Comment (jakirkham) [inline recipes/tesseract/meta.yaml]: Just as you nicely did above, please add...

```yaml
  build:
    - {{ compiler('c') }}
```

Then drop both `gcc` and `llvm` from here.

File: recipes/tesseract/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    - libpng
+    - jpeg
+    - libtiff
+    - automake
+    - autoconf
+    - libtool
+    - zlib
+    - leptonica
+  run:
+    - libgcc    # [linux]
```
Comment (jakirkham) [inline recipes/tesseract/meta.yaml]: This is automatically added by the `compiler` syntax noted above. So we don't need to add it explicitly. IOW please go ahead and drop this.

File: recipes/tesseract/build.sh [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+./autogen.sh
```
Comment (jakirkham) [inline recipes/tesseract/build.sh]: Normally people run `autoreconf -fi`, which does this and some other things. Also it appears there is a `CMakeList.txt`. So you could use `cmake` instead if you rather. The latter is a bit easier IMHO to use on Windows, but works fine on Unix too.

**Comment by jakirkham** [comment]
HTH :)

**Comment by jakirkham** [comment]
Raised issue ( https://github.com/conda/conda-build/issues/2921 ) on the `conda-build` tracker. Let's see what they say.

**Comment by msarahan** [comment]
I'm not certain, but this may be a bug in conda-build where the collection of variables is shared across more builds than it should be.  If the first recipe doesn't use openjpeg and cuts that variable out, then the second one does use it and doesn't find it.  Since conda-build doesn't actually delete variables from the variant set, I think this is a bug in staged-recipes scripts.  In particular, this single config object https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/build_all.py#L48-L59 should probably be copied in the loop at https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/compute_build_graph.py#L391

A workaround would be to submit recipes separately.

**Comment by CurtLH** [comment]
Thanks for the response @msarahan!  

@jakirkham -- I'll go ahead and split the current pull-request into two separate pull-requests.

**Comment by CurtLH** [comment]
Here is the separate [PR for leptonica](https://github.com/conda-forge/staged-recipes/pull/5970).

However, it seems that the PR for Leptonica is now failing because it [cannot find `openjpeg`](https://github.com/conda-forge/staged-recipes/pull/5970#issuecomment-392957627), even though it is specified in both `run` and `host`.

**Comment by mbargull** [comment]
This is clearly a non-bio package and thus belongs here. We'll remove its recipe from Bioconda.

**Comment by mbargull** [comment]
@CurtLH, since `leptonica` is now on the channel, you should be able to trigger a new CI run with something like
```
git commit --allow-empty -m trigger && git push
```

<sub>(PS: you :-1:'ed my previous comment :open_mouth: :broken_heart: ... :stuck_out_tongue_winking_eye: )</sub>

**Comment by scopatz** [comment]
> git commit --allow-empty -m trigger && git push

It would be really nice to put that in the conda-forge docs

**Comment by CurtLH** [comment]
Thanks for triggering the build @scopatz!

I'm not very (or at all) experienced in debugging these types of recipes, but it looks like this has something to do with why it failed:

```
./configure: line 17577: syntax error near unexpected token `LEPTONICA,'
./configure: line 17577: `PKG_CHECK_MODULES(LEPTONICA, lept >= 1.74, have_lept=true, have_lept=false)'
```

Do you have any thoughts on what to do?

**Comment by scopatz** [comment]
Did you push up an empty commit to trigger a rebuild? Also I think you need to wait for the leotonica package to actually appear on anaconda.org

**Comment by CurtLH** [comment]
Leptonica is now in [Anaconda cloud](https://anaconda.org/conda-forge/leptonica), and I just pushed an empty commit, but still the same error.

**Comment by mbargull** [comment]
Probably just needs `pkg-config` as a `build` dependency, see https://github.com/tesseract-ocr/tesseract/issues/941.

File: recipes/tesseract/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - libpng
+    - jpeg
+    - libtiff
+    - automake
+    - autoconf
```
Comment (mbargull) [inline recipes/tesseract/meta.yaml]: `automake` and `autoconf` should go into `requirements/build`, if possible.

**Comment by CurtLH** [comment]
This one is ready for review @conda-forge/staged-recipes!

**Comment by isuruf** [comment]
Can you fix these warnings,
```
WARNING (tesseract,bin/tesseract): Needed DSO lib/libwebp.so.7 found in ['libwebp']
WARNING (tesseract,bin/tesseract): .. but ['libwebp'] not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
WARNING (tesseract,bin/tesseract): Needed DSO lib/liblzma.so.5 found in ['xz']
WARNING (tesseract,bin/tesseract): .. but ['xz'] not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
WARNING (tesseract,bin/tesseract): Needed DSO lib/libopenjp2.so.7 found in ['openjpeg']
WARNING (tesseract,bin/tesseract): .. but ['openjpeg'] not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
WARNING (tesseract,bin/tesseract): Needed DSO lib/libgif.so.7 found in ['giflib']
WARNING (tesseract,bin/tesseract): .. but ['giflib'] not in reqs/run, i.e. it is overlinked (likely) or a missing dependency (less likely)
```

**Comment by CurtLH** [comment]
Thanks for catching that @isuruf.  Would this be the right way to fix this issue?

1. Wait to `tesseract` feedstock to be created
2. Fork `tesseract` feedstock
3. Add `libwebp`, `xz`, `openjpeg`, and `giflib` under `run:requirements`
4. Increment build number to `1`
5. Merge fork into master

**Comment by scopatz** [comment]
Yes, exactly. But you should issue a PR into the feedstock from your fork.

**Comment by CurtLH** [comment]
That's what I meant.  Will do.  Thank you for your help!

**Comment by scopatz** [comment]
Thanks!

**Comment by isuruf** [comment]
Remember to do a `conda smithy rerender` after adding the requirements.

**Comment by isuruf** [comment]
You'll also want to do this for leptonica. See https://travis-ci.org/conda-forge/leptonica-feedstock/jobs/386797659#L1862-L1865

**Comment by CurtLH** [comment]
Will do.  Thanks for spotting that!

**Pull request merged**